File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.cc
#include <bits/stdc++.h>

using namespace std;

int n;
int arr[1001];
int dp[1001] = { 0, };
int result = 0;

int main()
{
	cin >> n;

	for (int i = 0; i < n; i++)
	{
		cin >> arr[i];
	}

	for (int i = 0; i < n; i++)
	{
		int Max = 0;

		for (int j = 0; j < i; j++)
		{
			if (arr[i] > arr[j])
			{
				if (Max < dp[j])
					Max = dp[j];
			}
		}
		dp[i] = Max +1;

		result = max(dp[i], result);
	}

	cout << result;

	return 0;
}

File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.java
import java.util.*;
import java.io.*;

public class Main {
	
	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static StringTokenizer st;
	
	static int n, m, Max;
	static int[] arr, result;
	
	public static void main(String[] args) throws IOException {
		n = Integer.parseInt(br.readLine());
		
		arr = new int[n];
		result = new int[n];
		
		st = new StringTokenizer(br.readLine());
		
		for (int i = 0 ; i < n ; i ++) {
			arr[i] = Integer.parseInt(st.nextToken());
			result[i] = 1;
		}
		
		
		
		// n^2의 풀이
		
		for (int i = 1 ; i < n ; i++) {
			for (int j = 0 ; j < i ; j ++) {
				if (arr[j] < arr[i])
				{
					result[i] = Math.max(result[i], result[j] + 1);
				}
			}
		}
		
		for (int i = 0 ; i < n ; i ++) {
			Max = Math.max(result[i], Max);
		}
		
		System.out.println(Max);
		
	}
	
}